encoding.go

```diff
@@ -0,0 +1,189 @@
+type Serializer interface {
+	// Encoding returns the encoding for this serializer.
+	Encoding() Encoding
+
+	// Request creates a transport.Request from the given []byte input.
+	Request(body []byte) (*transport.Request, error)
+
+	// Response converts a transport.Response into something that can be displayed to a user.
+	// For non-raw encodings, this is typically a map[string]interface{}.
+	Response(body *transport.Response) (interface{}, error)
+
+	// IsSuccess returns whether the given body is considered a successful response.
+	IsSuccess(body *transport.Response) error
+}
+
+// The list of supported encodings.
+const (
+	UnspecifiedEncoding Encoding = ""
+	JSON                Encoding = "json"
+	Thrift              Encoding = "thrift"
+	Raw                 Encoding = "raw"
+)
+
+var (
+	errNilEncoding          = errors.New("cannot Unmarshal into nil Encoding")
+	errUnrecognizedEncoding = errors.New("unrecognized encoding, must be one of: json, thrift, raw")
+	errHealthThriftOnly     = errors.New("--health can only be used with Thrift")
+	errMissingMethodName    = errors.New("no method specified, specify --method method")
+)
+
+func (e Encoding) String() string {
+	return string(e)
+}
+
+// UnmarshalText imlements the encoding.TextUnmarshaler interface used by JSON, YAMl, etc.
+func (e *Encoding) UnmarshalText(text []byte) error {
+	if e == nil {
+		return errNilEncoding
+	}
+
+	switch s := strings.ToLower(string(text)); s {
+	case "", "json", "thrift", "raw":
+		*e = Encoding(s)
+		return nil
+	default:
+		return errUnrecognizedEncoding
+	}
+}
+
+// UnmarshalFlag allows Encoding to be used as a flag.
+func (e *Encoding) UnmarshalFlag(s string) error {
+	return e.UnmarshalText([]byte(s))
+}
+
+func (e Encoding) supportsHealth() bool {
+	switch e {
+	case UnspecifiedEncoding, Thrift:
+		return true
+	default:
+		return false
+	}
+}
+
+// NewSerializer creates a Serializer for the specific encoding.
+func NewSerializer(opts RequestOptions) (Serializer, error) {
+	e := opts.Encoding
+
+	if opts.Health {
+		if !e.supportsHealth() {
+			return nil, errHealthThriftOnly
+		}
+		if opts.MethodName != "" {
+			return nil, errHealthAndMethod
+		}
+
+		method, spec := getHealthSpec()
+		return thriftEncoding{method, spec}, nil
+	}
+
+	if opts.MethodName == "" {
+		return nil, errMissingMethodName
+	}
+
+	if e == UnspecifiedEncoding && strings.Contains(opts.MethodName, "::") {
+		e = Thrift
+	}
+
+	switch e {
+	case Thrift:
+		return newThriftEncoding(opts.ThriftFile, opts.MethodName)
+	case JSON:
+		return jsonEncoding{opts.MethodName}, nil
+	case Raw:
+		return rawEncoding{opts.MethodName}, nil
+	}
+
+	return nil, errUnrecognizedEncoding
+}
+
+type jsonEncoding struct {
+	methodName string
+}
+
+func (e jsonEncoding) Encoding() Encoding {
+	return JSON
+}
+
+func (e jsonEncoding) Request(input []byte) (*transport.Request, error) {
+	data, err := unmarshalJSONInput(input)
+	if err != nil {
+		return nil, err
+	}
+
+	bs, err := json.Marshal(data)
```

Comment:
Yeah just checking for valid JSON, and to send the same consistent JSON request (e.g. the ordering of fields etc is deterministic, whitespace is trimmed).

I'll add a comment.